Offline Mode: Milestone 2

kean opened the pull request:
This PR contains the changes from the second milestone.

Planned changes:

- https://github.com/wordpress-mobile/WordPress-iOS/pull/22848
- [x] Rework the publishPost method and editor actions to make it easier to remove the deprecated code later
- [x] Integrate PostRepository.save in the draft saving flow
- [x] Integrate PostRepository.save in the remaining flows: delete, restore, etc
- [x] Implement the conflict resolution dialog

staskus committed:
Increase number of weeks to fetch for monthly Stats Traffic chart

kean committed:
Integrate PostRepository.save in publishing flow

kean committed:
Remove redundant date parameter

kean committed:
Improve error handling for PostRepository.save

kean committed:
Show Done instead of Close in Post Settings screen

kean committed:
Integrate PostRepository.save in Post Settings

kean committed:
Mark post as synced after upload

staskus committed:
Update existing Pods

kean committed:
Remove outdated comment

dangermattic commented:
<!--
  0 Errors
  1 Warning: PR is not assigned to a milest...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is not assigned to a milestone.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

kean committed:
Fix applying changes without creating a new revision

kean committed:
Close the editor after post is deleted

kean committed:
Update

kean committed:
Offline Mode: Integrate new PostRepository.save method in main publishing flows (#22848)

dvdchr committed:
Fix blinking navigation bar on dark mode in Reader Post

kean commented:
Hey, @momo-ozawa. I disabled `.syncPublishing` for now to make this mergable.

dangermattic commented:
<!--
  0 Errors
  2 Warnings: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

Commit:
add conflict resolution view

Commit:
show conflict resolution dialog

Commit:
fix typo

kean committed:
Merge branch 'trunk' into feature/offline-mode-milesone-2

kean committed:
Update podfile

kean committed:
Implement moveToDraft using new APIs

kean committed:
Add new sync engine for draft posts

kean committed:
Remove secondary actions from the editor

kean committed:
Integrate publishing for drafts with the sync engine

kean committed:
Remove post.status setter usage

kean committed:
Update Post Settings to use the new APIs

kean committed:
Offline Mode: Integrate new APIs in moveToDraft (#22892)

kean committed:
Offline Mode: Sync Engine (#22899)

kean committed:
Implement trashing using new APIs

kean committed:
Remove unneeded restore method

kean committed:
Implement permanent deletion

Commit:
rename version enum

kean committed:
Fix an issue with changing publish date and status

kean committed:
Fix typos

kean committed:
Update checkmark style

kean committed:
Add support for pending posts

kean committed:
Update context menu actions for pending posts

kean committed:
Allow saving pending posts

kean committed:
Fix false positive in data conflict

kean committed:
Fix remaining issues

kean committed:
Match the meta header color with the rest of the content during the upload state

Commit:
save selected version

Commit:
show error alert if conflict resolution fails

Commit:
notify editor post conflict has been resolved

kean committed:
Update unit tests

kean committed:
Move hasUnsyncedChanges to PostRepository

kean committed:
Implement improved autosave dialog

kean committed:
Update support for publishing pages

kean committed:
Remove the local-changes badge from pages

kean committed:
Update View button to no longer depend on remoteStatus

kean committed:
Integrate PostSyncStateViewModel in the page list

kean committed:
Remove blank page notice

kean committed:
Implement sync state updates for PageListCell

kean committed:
Simplify query for post statuses

kean committed:
Udpate set parent page feature to use the new endpoints

kean committed:
Offline Mode: Remove secondary actions from the editor (#22913)

kean committed:
Remove ViewController-based tests with no mocks

kean committed:
Add feature flag for hasChanges change

Commit:
ensure single selection for post version

kean committed:
Update UI tests to not use the new SyncPublishing feature flag

kean committed:
Offline Mode: Remove post.status usage from the publishing flows (#22930)

kean committed:
Offline Mode: Publishing Drafts (#22919)

kean committed:
Offline Mode: Update Post Settings to use the new APIs (#22934)

kean committed:
Fix an issue with posts deleted on the remote not being deleted locally

Commit:
disable nav bar buttons while saving

Commit:
extract toolbar content

Commit:
disable interactive dismiss

kean committed:
Add testInitializeSync

kean committed:
PostListCell to update itself when sync state changes

Commit:
remove unnecessary code

kean committed:
Add crash protection

kean committed:
Offline Mode: Trash, Delete, Restore (#22947)

kean committed:
Offline Mode: Add support for pending posts (#22949)

kean committed:
Add support for using PostCoordinator.save for drafts

kean committed:
Update PostRepository.sync to work offline when changes are fully reverted locally

kean committed:
Update preview support to use new APIs

kean committed:
Add revisions support

kean committed:
Add support for post metadata

kean committed:
Offline Mode: Pages Support (#22964)

kean committed:
Revert the list cell changes

kean committed:
Increase the retry delay

kean committed:
Add WPAssert and error tracking

kean committed:
Fix the syncPublishing FF use in PostCoordinator

Commit:
track events for conflict resolution dialog

Commit:
undo pbxproj change

kean committed:
Offline Mode: Autosave and State Restoration (#22963)

kean committed:
Offline Mode: Simplify Queries in Post List (#22966)

kean committed:
Update WordPress/WordPressTest/PostCoordinatorSyncTests.swift

Co-authored-by: Momo Ozawa <momo-ozawa@users.noreply.github.com>

kean committed:
Update WordPress/WordPressTest/PostCoordinatorSyncTests.swift

Co-authored-by: Momo Ozawa <momo-ozawa@users.noreply.github.com>

kean committed:
Offline Mode: Previews (#22973)

kean committed:
Update WordPress/Classes/ViewRelated/Post/PostEditor+MoreOptions.swift

Co-authored-by: Momo Ozawa <momo-ozawa@users.noreply.github.com>

kean committed:
Fix unit tests

kean committed:
Offline Mode: Revisions (#22974)

kean committed:
Offline Mode: Metadata (#22976)

kean committed:
Improve migration to the new system

kean committed:
Merge branch 'trunk' into feature/offline-mode-milesone-2

kean committed:
Cleanup

kean committed:
Fix post settings (prod version)

kean committed:
Fix unit tests

momo-ozawa committed:
Offline Mode: Add conflict resolution dialog (#22879)

Commit:
show conflict resolution dialog from post list

kean committed:
Fix cell reload on postCoordinatorDidUpdate

kean committed:
Fix trash (XMLRPC issue)

kean committed:
Offline Mode: Migration (#22979)

kean committed:
Update propertu names

Commit:
subscribe to resolve conflict notif in PostEditor

kean committed:
Offline Mode: Instrumentation (#22980)

kean committed:
Merge branch 'trunk' into feature/offline-mode-milesone-2

Commit:
resolve conflict on post.original()

Commit:
use postCoordinatorDidUpdate notif to update editor

Commit:
fire notifications using pre-existing method

kean committed:
Disable context menu when post is not editable

Commit:
subscribe to postCoordinatorDidUpdate notif in prepublishing vc

Commit:
remove the postConflictResolved notification

conflict resolution now relies on the postCoordinatorDidUpdate notification

Commit:
rely on postID to identify updated post

kean committed:
Fix discard revision

kean committed:
Fix an issue with PostCoordinator always sending a password field

kean committed:
Offline Mode: Fix cell reload on postCoordinatorDidUpdate (#22986)

Commit:
fix swiftlint error

Commit:
use wpAssert

kean committed:
Update WordPress/Classes/Services/PostRepository.swift

Co-authored-by: Momo Ozawa <momo-ozawa@users.noreply.github.com>

kean committed:
Offline Mode: Disable context menu when post is not editable (#22993)

kean committed:
Offline Mode: Fix discard changes (#22996)

kean committed:
Fix an issue with Post Settings Save button not being updated

kean committed:
Remove Publish Date option from Post Settings for draft and pending posts

kean committed:
Update comments

kean committed:
Remove schedule action from PostEditorState

kean committed:
Add PrepublishingViewModel

kean committed:
Add DeprecatedPrepublishingViewController

kean committed:
Rework the prepublishing sheet

kean committed:
Offline Mode: Fix trashing (XMLRPC) (#22987)

Commit:
check for specific error cases when opting out of retry

Commit:
send notification after resetting error state

Commit:
Revert "remove the postConflictResolved notification"

This reverts commit b5a97ce988b5b2981f446c9a949ae82199983600.

Commit:
use postConflictResolved notif to update editor

Intentionally not using postCoordinatorDidUpdate to avoid breaking the sync mechanism.

The postCoordinatorDidUpdate is called frequently during the upload. If we use the postCoordinatorDidUpdate here, the editor will keep creating revisions when it doesn't need to, and it will break the hasChanges mechanism.

Commit:
show resolve conflict dialog for post list, page list, and search

Commit:
clear conflict error by calling startSync

Commit:
add clarifying comment for resolve conflict notification

kean committed:
Fix an issue with trashing a permanently deleted post

kean committed:
Fix publish date formatting when .syncPublishing is off

kean committed:
Improve WPInstrumentation

kean committed:
Add more assertions

kean committed:
Add additional PostRepository analytics

kean committed:
Fixes for scheduled post handling

kean committed:
Fix media upload tracking

kean committed:
Fix an issue with update button not being available immediatelly

momo-ozawa committed:
Offline mode: Show conflict resolution dialog from post list (#22992)

kean committed:
Offline Mode: Fix update button not being enabled after changing settings (#23021)

kean committed:
Integrate if_not_modified_since for XMLRPC

kean committed:
Offline Mode: Fix media upload status (#23017)

kean committed:
Merge branch 'feature/offline-mode-milesone-2' into task/offline-mode-fix-publish-date-handling

kean committed:
Fix race condition in EditorGutenbergTests

kean committed:
Offline Mode: Prepublishing sheet fixes (#22998)

kean committed:
Offline Mode: Fix an issue with trashing a permanently deleted posts (#23004)

kean committed:
Revert the draft udpate fix as it is not needed for wp.editPost

kean committed:
Offline Mode: Fix a couple of issues with scheduled posts (XMLRPC) (#23016)

kean committed:
Add assertion sampling

kean committed:
Offline Mode: Analycs improvements (#23011)

kean committed:
Merge branch 'trunk' into feature/offline-mode-milesone-2

kean committed:
Offline Mode: Assertion Sampling (#23024)

kean committed:
Offline Mode: Add if_not_modified_since support for self-hosted sites (#23022)

kean committed:
Fix an issue with posting a re-drafted scheduled post

kean committed:
Offline Mode: Fix an issue with posting a re-drafted scheduled post (#23028)

Commit:
use new delete API for swipe action

momo-ozawa committed:
Offline Mode: Fix trash post swipe action (#23029)

kean committed:
Merge branch 'trunk' into feature/offline-mode-milesone-2
